File: solutions/scala/protein-translation/1/src/main/scala/ProteinTranslation.scala
object ProteinTranslation {
  private val Stop = "STOP"

  private val proteinsByCodon = Map(
    "AUG" -> "Methionine",
    "UUU" -> "Phenylalanine",
    "UUC" -> "Phenylalanine",
    "UUA" -> "Leucine",
    "UUG" -> "Leucine",
    "UCU" -> "Serine",
    "UCC" -> "Serine",
    "UCA" -> "Serine",
    "UCG" -> "Serine",
    "UAU" -> "Tyrosine",
    "UAC" -> "Tyrosine",
    "UGU" -> "Cysteine",
    "UGC" -> "Cysteine",
    "UGG" -> "Tryptophan",
    "UAA" -> Stop,
    "UAG" -> Stop,
    "UGA" -> Stop
  )

  def proteins(rna: String): Seq[String] =
    rna
      .grouped(3)
      .map(proteinsByCodon)
      .takeWhile(_ != Stop)
      .toSeq
}

File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let private translateCodon = function
    | "AUG" -> Some "Methionine"
    | "UUU" | "UUC" -> Some "Phenylalanine"
    | "UUA" | "UUG" -> Some "Leucine"
    | "UCU" | "UCC" | "UCA" | "UCG" -> Some "Serine"
    | "UAU" | "UAC" -> Some "Tyrosine"
    | "UGU" | "UGC" -> Some "Cysteine"
    | "UGG" -> Some "Tryptophan"
    | "UAA" | "UAG" | "UGA" -> None
    | _ -> None

let proteins (rna: string): string list =
    rna
    |> Seq.chunkBySize 3
    |> Seq.map (fun codon -> System.String codon)
    |> Seq.map translateCodon
    |> Seq.takeWhile Option.isSome
    |> Seq.choose id
    |> Seq.toList
